lightning/src/ln/channelmanager.rs

```diff
@@ -4069,135 +4098,136 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
+
+			claimable_amt_msat += htlc.value;
+		}
+		if sources.is_empty() || expected_amt_msat.is_none() {
+			self.pending_claimed_payments.lock().unwrap().remove(&payment_hash);
+			log_info!(self.logger, "Attempted to claim an incomplete payment which no longer had any available HTLCs!");
+			return;
+		}
+		if claimable_amt_msat != expected_amt_msat.unwrap() {
+			self.pending_claimed_payments.lock().unwrap().remove(&payment_hash);
+			log_info!(self.logger, "Attempted to claim an incomplete payment, expected {} msat, had {} available to claim.",
+				expected_amt_msat.unwrap(), claimable_amt_msat);
+			return;
+		}
+		if valid_mpp {
+			for htlc in sources.drain(..) {
+				if channel_state.is_none() { channel_state = Some(self.channel_state.lock().unwrap()); }
+				if let Err((pk, err)) = self.claim_funds_from_hop(channel_state.take().unwrap(), htlc.prev_hop,
+					payment_preimage,
+					|_| Some(MonitorUpdateCompletionAction::PaymentClaimed { payment_hash }))
+				{
+					if let msgs::ErrorAction::IgnoreError = err.err.action {
+						// We got a temporary failure updating monitor, but will claim the
+						// HTLC when the monitor updating is restored (or on chain).
+						log_error!(self.logger, "Temporary failure claiming HTLC, treating as success: {}", err.err.err);
+					} else { errs.push((pk, err)); }
 				}
 			}
-
-			if claimed_any_htlcs {
-				self.pending_events.lock().unwrap().push(events::Event::PaymentClaimed {
-					receiver_node_id,
-					payment_hash,
-					purpose: payment_purpose,
-					amount_msat: claimable_amt_msat,
-				});
+		}
+		mem::drop(channel_state);
+		if !valid_mpp {
+			for htlc in sources.drain(..) {
+				let mut htlc_msat_height_data = byte_utils::be64_to_array(htlc.value).to_vec();
+				htlc_msat_height_data.extend_from_slice(&byte_utils::be32_to_array(
+					self.best_block.read().unwrap().height()));
+				self.fail_htlc_backwards_internal(
+					HTLCSource::PreviousHopData(htlc.prev_hop), &payment_hash,
+					HTLCFailReason::Reason { failure_code: 0x4000|15, data: htlc_msat_height_data },
+					HTLCDestination::FailedPayment { payment_hash } );
 			}
+		}
 
-			// Now we can handle any errors which were generated.
-			for (counterparty_node_id, err) in errs.drain(..) {
-				let res: Result<(), _> = Err(err);
-				let _ = handle_error!(self, res, counterparty_node_id);
-			}
+		// Now we can handle any errors which were generated.
+		for (counterparty_node_id, err) in errs.drain(..) {
+			let res: Result<(), _> = Err(err);
+			let _ = handle_error!(self, res, counterparty_node_id);
 		}
 	}
 
-	fn claim_funds_from_hop(&self, channel_state_lock: &mut MutexGuard<ChannelHolder<<K::Target as KeysInterface>::Signer>>, prev_hop: HTLCPreviousHopData, payment_preimage: PaymentPreimage) -> ClaimFundsFromHop {
+	fn claim_funds_from_hop<ComplFunc: FnOnce(Option<u64>) -> Option<MonitorUpdateCompletionAction>>(&self,
+		mut channel_state_lock: MutexGuard<ChannelHolder<<K::Target as KeysInterface>::Signer>>,
```

Comment:
Hmm I'm trying to understand this, so apologies if my question is dumb:

Now that we're retaking the `channel_state_lock` in `claim_funds`, is the only reason we're passing around the `channel_state_lock` to `claim_funds_from_hop` and `claim_funds_internal`, that we'd like to ensure that hold the `channel_state_lock` (ie. not dropping the channel) until completing `claim_funds_from_hop` after updating the channel by calling `Channel::update_fulfill_htlc` in `ChannelManager::internal_update_fulfill_htlc`?
And if that hadn't been the case, we could have avoided passing around the `channel_state_lock`, and just acquired it in the beginning of `claim_funds_from_hop`.

Is there any way we can achieve not having to pass it around, as it would be a pretty nice cleanup?

Reply:
Ok :). I'll let you determine once I've finished updating #1507 if you believe it's worth it then.